middle_end/flambda/to_cmm/un_cps_closure.ml

```diff
@@ -105,40 +113,73 @@ let order_env_vars env l acc =
             Var_within_closure.print env_var
     ) acc l
 
-let order env closures env_vars =
-  order_env_vars env env_vars
-    (order_closures env closures Numbers.Int.Map.empty)
-
-let rec layout_aux map acc i =
-  match Numbers.Int.Map.find_first (fun j -> j >= i) map with
-  | exception Not_found -> List.rev acc
-  | (j, (Closure _ as slot)) when j = 0 ->
-      assert (i = j);
-      assert (acc = []);
-      layout_aux map [(0, slot)] (i + 1)
-  | (j, (Closure _ as slot)) ->
-      layout_aux map ((j, slot) :: (j - 1, Infix_header) :: acc) (j + 1)
-  | (j, (Env_var _ as slot)) ->
-      layout_aux map ((j, slot) :: acc) (j + 1)
-  | (_, Infix_header) ->
+let layout_aux j slot (startenv, acc_slots) =
+  match slot with
+  | Closure _ when j = 0 ->
+    assert (acc_slots = []);
+    assert (startenv = None);
+    let acc_slots = [(0, slot)] in
+    startenv, acc_slots
+  | Closure _ ->
+    assert (startenv = None);
```

Comment:
You should put a comment here explaining why this assertion holds (this is a second check that closures always occur before environment variables, if this doesn't hold this should have already been caught earlier and reported through a fatal error)

Reply:
Comment added